walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,54 +15,162 @@
+        // Don't optimize this because optimization should only apply to
+        // mandatory attendees 
+        int[] meetingStatesWithAllAttendees = 
+            getMinutesState(events, request.getAllAttendees());
+        Collection<TimeRange> meetingTimeRangesWithAllAttendees = 
+            convertStateToTimeRanges(meetingStatesWithAllAttendees, duration, request.getAllAttendees().size(), false);
 
-        // Get all the event time ranges and sort them by the start time. 
-        // Only add to the list if at least one attendee from the event is 
-        //a required attendee in the meeting request. (Otherwise, the event is irrelevant.)
+        // If there are no time ranges that work for all attendees, get the optimized time
+        // ranges that work for best for the mandatory attendees.
+        // If there are no mandatory attendees, treat the optional attendees as mandatory attendees
+        // and optimize the timeranges
+        if (request.getAttendees().size() == 0){
+            Collection<TimeRange> meetingTimeRangesWithOptionalAttendees = 
+                convertStateToTimeRanges(meetingStatesWithAllAttendees, duration, request.getAllAttendees().size(), true);
+            return meetingTimeRangesWithOptionalAttendees;
+        }else if (meetingTimeRangesWithAllAttendees.size() == 0) {
+            int[] meetingStatesWithMandatoryAttendees = 
+                getMinutesState(events, request.getAttendees());
+            Collection<TimeRange> meetingTimeRangesWithMandatoryAttendees = 
+                convertStateToTimeRanges(meetingStatesWithMandatoryAttendees, duration, request.getAttendees().size(), true);
+            return meetingTimeRangesWithMandatoryAttendees;
+        }else{
+            return meetingTimeRangesWithAllAttendees;
+        }
+    }
+
+    // Get all the valid event time ranges and sort them by the start time. 
+    // Only add to the list if at least one attendee from the event is 
+    // a required attendee in the meeting request. (Otherwise, the event is irrelevant.)
+    public List<TimeRange> getValidEventTimeRanges(Collection<Event> events, Collection<String> requestAttendees) {
         List<TimeRange> eventTimeRanges = events.stream().filter(
-         event -> event.getAttendees().stream().anyMatch(
-                attendee -> request.getAttendees().contains(attendee)))
-                .map(e -> e.getWhen())
-                .collect(Collectors.toList());
+            event -> event.getAttendees().stream().anyMatch(
+            attendee -> requestAttendees.contains(attendee)))
+            .map(e -> e.getWhen())
+            .collect(Collectors.toList());
         Collections.sort(eventTimeRanges, TimeRange.ORDER_BY_START);
+        return eventTimeRanges;
+    }
+
+    // Function that returns all valid time ranges based on a list of event time ranges
+    // Approach 1: start form start of day, and add available timeranges as we traverse through the events' timeranges
```

Comment:
Do the 2 approaches achieve the same goal? Is this a requirement from the challenge? (we typically don't leave multiple solutions to the same problem in because it can be confusing for readers in the future)

Reply:
Okay! I just removed them.